### PR TITLE
SameSite=None Incompatible Clients implementation and FAT Tests

### LINKED ISSUE #19887: Add FAT Testing for Incompatible 'SameSite=None' Clients
for #14842

### LINKED ISSUE #24901: SameSite=None known incompatible clients
This change takes into consideration the known incompatible clients, in order to remove - when applicable - the SameSite attribute if set to SameSite=None. This is modeled after the following Chromium update: https://www.chromium.org/updates/same-site/incompatible-clients

Addresses the feature: https://github.com/OpenLiberty/open-liberty/issues/14842

No changes are required from Session, Security or application defined cookies. The User-Agent request header is evaluated for each cookie and|or Set-Cookie header before the final Set-Cookie header values are written out to the underlying socket connection. If the User-Agent is found to be unsupported, and the SameSite cookie attribute is defined with a value of SameSite=None, the attribute is removed.

### PR DESCRIPTION
**Pull request by tssmith04**
Issue #19887

### CODE REVIEW

**Comment by tssmith04** [comment]
#build

File: dev/com.ibm.ws.webcontainer.servlet.4.0_fat.2/fat/src/com/ibm/ws/fat/wc/FATSuite.java [hunk 1]
```diff
@@ -47,11 +48,12 @@
...
-                WCSameSiteCookieAttributeTests.class,
-                WCSameSiteCookieAttributeSecurityTest.class,
-                WCPartitionedAttributeTests.class,
-                WCPartitionedCookieAttributeSecurityTest.class
+                // WCResponseHeadersTest.class,
+                // WCSameSiteCookieAttributeTests.class,
+                // WCSameSiteCookieAttributeSecurityTest.class,
+                // WCPartitionedAttributeTests.class,
+                // WCPartitionedCookieAttributeSecurityTest.class,
+                WCSameSiteIncompatibleClientsTests.class
```
Comment (volosied) [inline dev/com.ibm.ws.webcontainer.servlet.4.0_fat.2/fat/src/com/ibm/ws/fat/wc/FATSuite.java]: Uncomment

File: dev/com.ibm.ws.webcontainer.servlet.4.0_fat.2/publish/servers/servlet40_incompatible/server.xml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+                  httpsPort="${bvt.prop.HTTP_default.secure}">
+                  <samesite none="AddCookieCookie"/>
+    </httpEndpoint>
+
+    <javaPermission className="java.util.PropertyPermission"
+        name="org.glassfish.web.rfc2109_cookie_names_enforced"
+        actions="read"
+        codebase="${server.config.dir}dropins/IncompatibleClientTest.war"/>
+
+</server>
```
Comment (volosied) [inline dev/com.ibm.ws.webcontainer.servlet.4.0_fat.2/publish/servers/servlet40_incompatible/server.xml]: Can you add a new line

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpBaseMessageImpl.java [hunk 3]
```diff
@@ -2970,6 +2968,18 @@ else if (getServiceContext().getHttpConfig().onlySameSiteStar()) {
...
+            if (cookie.getAttribute("samesite") != null && cookie.getAttribute("samesite").equals(HttpConfigConstants.SameSite.NONE.getName())) {
+                String userAgent = getServiceContext().getRequest().getHeader(HttpHeaderKeys.HDR_USER_AGENT).asString();
+                if (userAgent != null && SameSiteCookieUtils.isSameSiteNoneIncompatible(userAgent)) {
+                    //TODO: do we remove Secure, probably should be retained.
+                    cookie.setAttribute("samesite", null);
+                    // Partitioned should only be included when SameSite=None, so if it is incompatible Partitioned should be removed as well
+                    if (partitionedValue != null) {
+                        cookie.setAttribute("partitioned", null);
+                    }
+                }
```
Comment (volosied) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpBaseMessageImpl.java]: I would add a beta guard here as Paul mentioned. 

Add `import com.ibm.ws.kernel.productinfo.ProductInfo;` 
And then add a `ProductInfo.getBetaEdition()` check 

Add a jvm.options to the server with "-Dcom.ibm.ws.beta.edition=true"

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpBaseMessageImpl.java [hunk 4]
```diff
@@ -2970,6 +2968,18 @@ else if (getServiceContext().getHttpConfig().onlySameSiteStar()) {
                 }
             }
 
+            if (cookie.getAttribute("samesite") != null && cookie.getAttribute("samesite").equals(HttpConfigConstants.SameSite.NONE.getName())) {
+                String userAgent = getServiceContext().getRequest().getHeader(HttpHeaderKeys.HDR_USER_AGENT).asString();
+                if (userAgent != null && SameSiteCookieUtils.isSameSiteNoneIncompatible(userAgent)) {
+                    //TODO: do we remove Secure, probably should be retained.
```
Comment (volosied) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpBaseMessageImpl.java:2976]: Should we add a log statement saying an incompatible client was found?

**Comment by tssmith04** [comment]
#build

File: dev/com.ibm.ws.webcontainer.servlet.4.0_fat.2/publish/servers/servlet40_incompatible/bootstrap.properties [hunk 5]
```diff
@@ -0,0 +1,15 @@
+###############################################################################
+# Copyright (c) 2020, 2024 IBM Corporation and others.
```
Comment (pnicolucci) [inline dev/com.ibm.ws.webcontainer.servlet.4.0_fat.2/publish/servers/servlet40_incompatible/bootstrap.properties]: Why a 2020 copyright header for a new file?

Comment (tssmith04) [reply dev/com.ibm.ws.webcontainer.servlet.4.0_fat.2/publish/servers/servlet40_incompatible/bootstrap.properties]: Was just copy pasted over. I'll delete it

**Comment by tssmith04** [comment]
#libby

File: dev/com.ibm.ws.transport.http/test/com/ibm/ws/http/channel/test/api/SameSiteCookieUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+            //Versions Chrome 67 forward support SameSite=None. Test a supported newer version.
+            //Expected to return false.
+            //Chrome 67
+            userAgent = "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.2526.73 Safari/537.36";
+            assertFalse(SameSiteCookieUtils.isSameSiteNoneIncompatible(userAgent));
+
+            //Test UC browser before 12.13.2
+            //Test UC browser on 12.13.2
+            //Test UC browser after 12.13.2
+            //test bad format of product
```
Comment (volosied) [inline dev/com.ibm.ws.transport.http/test/com/ibm/ws/http/channel/test/api/SameSiteCookieUtilsTest.java]: This comment looks like it can be removed since the UC tests are next.

However, I don't see a bad format test? Could you create one?

File: dev/com.ibm.ws.webcontainer.servlet.4.0_fat.2/test-applications/IncompatibleClientTest.war/src/incompatible/servlets/IncompatibleClientServlet.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+public class IncompatibleClientServlet extends HttpServlet {
+
+    private static final long serialVersionUID = 1L;
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        PrintWriter writer = response.getWriter();
+        writer.print("Welcome to the IncompatibleClientServlet!");
+
+        // TODO: Add more coverage with Session cookies
```
Comment (volosied) [inline dev/com.ibm.ws.webcontainer.servlet.4.0_fat.2/test-applications/IncompatibleClientTest.war/src/incompatible/servlets/IncompatibleClientServlet.java]: I would update this comment to say session cookie is configured via server.xml

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/cookies/SameSiteCookieUtils.java [hunk 8]
```diff
@@ -0,0 +1,271 @@
+/*******************************************************************************
+ * Copyright (c) 2021 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
```
Comment (volosied) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/cookies/SameSiteCookieUtils.java:9]: Maybe a question for @pnicolucci . 


This code in this file looks to based off the pseudocode example provided in https://www.chromium.org/updates/same-site/incompatible-clients/   (which is licensed as Apache-2.0) 

Should we update this to be an Apache 2.0 License or keep it as EPL 2.0?

File: dev/com.ibm.ws.webcontainer.servlet.4.0_fat.2/test-applications/IncompatibleClientTest.war/src/incompatible/servlets/IncompatibleClientServlet.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        request.getSession();
+        response.setHeader("set-cookie", "SetHeaderCookie; SameSite=None");
+        response.addHeader("set-cookie", "AddHeaderCookie; SameSite=None");
+        response.addCookie(new Cookie("AddCookieCookie", "sugar"));
+
+        // Have a base cookie that always gets sent
+        response.addHeader("set-cookie", "BasicCookie");
```
Comment (volosied) [inline dev/com.ibm.ws.webcontainer.servlet.4.0_fat.2/test-applications/IncompatibleClientTest.war/src/incompatible/servlets/IncompatibleClientServlet.java]: I would explictly say that SameSite is not added to this cookie.  The server.xml config only applies to jsession and the AddCookieCookie cookie.

**Comment by tssmith04** [comment]
#build

**Comment by tssmith04** [comment]
#build

**Comment by tssmith04** [comment]
#build

**Comment by tssmith04** [comment]
#build

**Comment by tssmith04** [comment]
#build

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/cookies/SameSiteCookieUtils.java (hunk 8 above)
Comment (tssmith04) [reply dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/cookies/SameSiteCookieUtils.java:9]: Vlad and I talked about this and a comment acknowledging that we pulled from the pseudocode should suffice.

**Review by volosied — APPROVED** [review]
LGTM. Glad this is finally out of the way.

**Review by pnicolucci — APPROVED** [review]

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpBaseMessageImpl.java [hunk 10]
```diff
@@ -2970,6 +2969,20 @@ else if (getServiceContext().getHttpConfig().onlySameSiteStar()) {
                 }
             }
 
+            // Must be in beta to check for SameSite=None Incompatible clients
+            if (ProductInfo.getBetaEdition() && cookie.getAttribute("samesite") != null && cookie.getAttribute("samesite").equals(HttpConfigConstants.SameSite.NONE.getName())) {
+                String userAgent = getServiceContext().getRequest().getHeader(HttpHeaderKeys.HDR_USER_AGENT).asString();
+                if (userAgent != null && SameSiteCookieUtils.isSameSiteNoneIncompatible(userAgent)) {
+                    //TODO: do we remove Secure, probably should be retained.
+                    Tr.debug(tc, "Incompatible client for SameSite=None found with the following User-Agent: " + userAgent);
```
Comment (mrsaldana) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpBaseMessageImpl.java:2977]: Tr.debug statements are more performant when wrapped around the: 

`if(TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()`

Comment (volosied) [reply dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpBaseMessageImpl.java:2977]: Thanks for spotting that -- I created a PR here: https://github.com/OpenLiberty/open-liberty/pull/29506
